string ip
uint32 quality
